### PR TITLE
  **Breaking:** CON-1155 Improve Tree

### PR DESCRIPTION
**Pull request by mpotomin**
# Why breaking:
Changes the existing API of the Tree component by renaming `color` prop to `tagColor`. 
Motivation: to make API more explicit and clear, avoid implicitly confusing the prop name with the font color and other color props, that this PR adds. 

# Why draft:
- [ ] Apply updated tree in the UI and make sure all the cases follow the requirememtns
- [x] Improve selected state of the tree items
- [x] Make it possible to ignore search words highlight for selected items
- [x] Add custom right padding to be able to align row actions
- [x] Add handling for "loading" and "error" states
- [x] Make the tree able to handle "+ Add table" type of items. 

# Summary
This PR:
1. Fixes the jumps of the Accordion's section header when toggled
1. Extends the customisation options for the Tree component items:

# Related issue
https://contiamo.atlassian.net/browse/CON-1155

# To be tested

Me
- [x] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/Tree/TreeItem.tsx [hunk 1]
```diff
@@ -88,15 +102,25 @@ const highlightStyle: React.CSSProperties = {
...
   /* Split the label by caract properly and show the first line only */
   overflow-wrap: break-word;
   overflow: hidden;
   text-overflow: ellipsis;
   height: 16px;
 
-  font-size: ${({ theme }) => theme.font.size.small}px;
-  font-weight: ${({ theme, hasChildren }) => (hasChildren ? theme.font.weight.bold : theme.font.weight.medium)};
+  line-height: 16px;
+  color: ${({ fontColor }) => (fontColor ? expandColor(constants, fontColor) : "unset")};
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]: why `unset`?

File: src/Tree/ChildTree.tsx [hunk 2]
```diff
@@ -18,13 +18,15 @@ const Container = styled.div<{ hasChildren: boolean; disabled: boolean }>`
 `
 
 const ChildTree: React.SFC<Props> = ({
+  paddingLeft,
+  paddingRight,
   initiallyOpen,
   highlight,
   tag,
+  tagColor,
```
Comment (stereobooster) [inline src/Tree/ChildTree.tsx]: `**Breaking:**`?

Comment (mpotomin) [reply src/Tree/ChildTree.tsx]: Yes totally, good catch! 👍

File: src/Tree/TreeItem.tsx (hunk 1 above)
Comment (mpotomin) [reply src/Tree/TreeItem.tsx]: @stereobooster 
The idea is: if no `fontColor` provided explicitly - make the font color of the label same as the the one of parent - `<Header/>` in this case. 

Alternatively, I'd suggest using `inherit` here as a value. Would `inherit` convey the intent better in this case?

Comment (stereobooster) [reply src/Tree/TreeItem.tsx]: yes 👍

File: src/AccordionSection/AccordionSection.tsx [hunk 3]
```diff
@@ -29,7 +29,7 @@ const Container = styled.div<{ expanded: boolean }>`
 const Header = styled(SectionHeader)<{ expanded: boolean }>(({ theme, expanded }) => ({
   cursor: "pointer",
   borderTop: `1px solid ${theme.color.border.invisible}`,
-  borderBottom: expanded ? `1px solid ${theme.color.border.invisible}` : "none",
+  borderBottom: `1px solid ${expanded ? "transparent" : theme.color.border.invisible}`,
```
Comment (stereobooster) [inline src/AccordionSection/AccordionSection.tsx]: transparent vs invisible 🤔

Comment (mpotomin) [reply src/AccordionSection/AccordionSection.tsx]: Invisible is only a title 🤷‍♂, poor naming selection though, I agree. What would you suggest in this case @stereobooster?

**Comment by stereobooster** [comment]
Highlighted text makes non-emphasised lines to jump, for example type "Limited"

**Comment by stereobooster** [comment]
Can tab into disabled item

Screenshot 2020-02-21 at 16 47 51

File: src/Tree/TreeItem.tsx [hunk 4]
```diff
@@ -169,24 +206,36 @@ const TreeItem: React.SFC<TreeItemProps> = ({
...
+        strong={Boolean(strong)}
+        fontSize={fontSize ? fontSize : constants.font.size.small}
+        emphasized={Boolean(emphasized)}
+        monospace={Boolean(monospace)}
+      >
+        <Highlighter
+          textToHighlight={label}
+          highlightStyle={highlightStyle}
+          sanitize={sanitizeInput}
+          searchWords={!ignoreSearchWords ? searchWords : []}
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]: It maybe good idea to use `const emptyArray = []`, to get the same instance every time here

File: src/Tree/Tree.tsx [hunk 5]
```diff
@@ -7,19 +7,28 @@ import { IconComponentType } from "../Icon"
...
   iconColor?: string
   onClick?: (event: React.MouseEvent) => void
   onContextMenu?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
   cursor?: string
   actions?: React.ReactNode
   forwardRef?: (element?: HTMLElement | null) => any
+  strong?: boolean
+  fontSize?: number
+  fontColor?: string
+  emphasized?: boolean
```
Comment (stereobooster) [inline src/Tree/Tree.tsx]: `strong vs emphasized` 🤔

File: src/Tree/ChildTree.tsx [hunk 6]
```diff
@@ -33,11 +36,17 @@ const ChildTree: React.SFC<Props> = ({
...
   cursor,
   searchWords,
+  ignoreSearchWords,
   level,
   actions,
   hasIconOffset,
   onMouseEnter,
   onMouseLeave,
+  strong,
+  fontSize,
```
Comment (stereobooster) [inline src/Tree/ChildTree.tsx]: Does it make sense to change fontSize across tree nodes? Shouldn't they all be the same size 🤔

**Comment by stereobooster** [comment]
Generally looks good -  the only concern is "jumping" text on highlight in the tree

File: src/Tree/TreeItem.tsx (hunk 4 above)
Comment (mpotomin) [reply src/Tree/TreeItem.tsx]: Yep, makes sense!

File: src/Tree/ChildTree.tsx (hunk 6 above)
Comment (mpotomin) [reply src/Tree/ChildTree.tsx]: It is same size if not modified, but current designs in Zepplin require smaller sizes for items in the deeper levels of nesting. Let's make it tweakable so we have a flexibility to decide later 

Screenshot 2020-02-21 at 17 11 21

**Comment by mpotomin** [comment]
> Highlighted text makes non-emphasised lines to jump, for example type "Limited"

Let me tweak the highlight style a bit, maybe that can be fixed easily

**Pull request merged**